### PR TITLE
Add priority to hooks.

### PR DESCRIPTION
**Pull request by tobehardest**
🔍 What type of PR is this?
/kind feature

👀 What this PR does / why we need it:
Add priority to hooks.

🅰 Which issue(s) this PR fixes:
Fixes https://github.com/OpenIMSDK/OpenKF/issues/26

### CODE REVIEW

File: server/internal/middleware/hooks/url_trie/trie_test.go [hunk 1]
```diff
@@ -20,23 +20,46 @@ import (
...
 
+type Hook4 struct {
+	Hook
+}
+
+func (h Hook4) Priority() int64 {
+	return 2
+}
+
+func TestUrlTrie(t *testing.T) {
```
Comment (charles-chenzz) [inline server/internal/middleware/hooks/url_trie/trie_test.go:55]:
```
   type args struct {
      index int
   }
   tests := []struct {
      name string
      args args
      want string
   }{
      {name: "index=0", args: args{index: 0}, want: "Sunday"},
      {name: "index=1", args: args{index: 1}, want: "Monday"},
      {name: "index=2", args: args{index: 2}, want: "Tuesday"},
      {name: "index=3", args: args{index: 3}, want: "Wednesday"},
      {name: "index=4", args: args{index: 4}, want: "Thursday"},
      {name: "index=5", args: args{index: 5}, want: "Friday"},
      {name: "index=6", args: args{index: 6}, want: "Saturday"},
      {name: "index=-1", args: args{index: -1}, want: "Unknown"},
      {name: "index=8", args: args{index: 8}, want: "Unknown"},
   }
   for _, tt := range tests {
      t.Run(tt.name, func(t *testing.T) {
         if got := GetWeekDay(tt.args.index); got != tt.want {
            t.Errorf("GetWeekDay() = %v, want %v", got, tt.want)
         }
      })
   }
```


can you try to refactor to table driven like this?

**Comment by charles-chenzz** [comment]
image
and these commit can be squashed into one commit, I didn't see the reason that we need 3 commits

**Comment by IRONICBo** [comment]
Can you update code in `server/cmd/genhooks`? Which used to auto generate hook function.

**Review by charles-chenzz — COMMENTED** [review]
lgtm

**Review by AllianceTing — COMMENTED** [review]
approve to merge

**Comment by IRONICBo** [comment]
@tobehardest `Priority` need to be add in template. link: https://github.com/OpenIMSDK/OpenKF/blob/0751b16fd3598d5267e392b1441596353985af58/server/cmd/genhooks/pkg/template.go#L53

**Comment by Cathay-Chen** [comment]
@tobehardest If multiple routes are matched, the priority ranking is incorrect. Should the sorting be done at the location where the hook is executed?

**Pull request merged**